httpclient/config.go

```diff
@@ -0,0 +1,161 @@
+/*
+Copyright © 2024 Acronis International GmbH.
+
+Released under MIT license.
+*/
+
+package httpclient
+
+import (
+	"github.com/acronis/go-appkit/config"
+	"time"
+)
+
+const (
+	// DefaultLimit is the default limit for rate limiting.
+	DefaultLimit = 10
+
+	// DefaultBurst is the default burst for rate limiting.
+	DefaultBurst = 100
+
+	// DefaultWaitTimeout is the default wait timeout for rate limiting.
+	DefaultWaitTimeout = 10 * time.Second
+
+	// configuration properties
+	cfgKeyRetriesMax            = "retries.max_retry_attempts"
+	cfgKeyRateLimitsLimit       = "rate_limits.limit"
+	cfgKeyRateLimitsBurst       = "rate_limits.burst"
+	cfgKeyRateLimitsWaitTimeout = "rate_limits.wait_timeout"
+	cfgKeyTimeout               = "timeout"
+)
+
+// RateLimitConfig represents configuration options for HTTP client rate limits.
+type RateLimitConfig struct {
+	Limit       int           `mapstructure:"limit"`
+	Burst       int           `mapstructure:"burst"`
+	WaitTimeout time.Duration `mapstructure:"wait_timeout"`
+}
+
+// Set is part of config interface implementation.
+func (c *RateLimitConfig) Set(dp config.DataProvider) (err error) {
+	limit, err := dp.GetInt(cfgKeyRateLimitsLimit)
+	if err != nil {
+		return err
+	}
+	c.Limit = limit
+
+	burst, err := dp.GetInt(cfgKeyRateLimitsBurst)
+	if err != nil {
+		return nil
+	}
+	c.Burst = burst
+
+	waitTimeout, err := dp.GetDuration(cfgKeyRateLimitsWaitTimeout)
+	if err != nil {
+		return err
+	}
+	c.WaitTimeout = waitTimeout
+
+	return nil
+}
+
+// SetProviderDefaults is part of config interface implementation.
+func (c *RateLimitConfig) SetProviderDefaults(dp config.DataProvider) {
+	dp.SetDefault(cfgKeyRateLimitsLimit, DefaultLimit)
+	dp.SetDefault(cfgKeyRateLimitsBurst, DefaultBurst)
+	dp.SetDefault(cfgKeyRateLimitsWaitTimeout, DefaultWaitTimeout)
+}
+
+// TransportOpts returns transport options.
+func (c *RateLimitConfig) TransportOpts() RateLimitingRoundTripperOpts {
+	return RateLimitingRoundTripperOpts{
+		Burst:       c.Burst,
+		WaitTimeout: c.WaitTimeout,
+	}
+}
+
+// RetriesConfig represents configuration options for HTTP client retries policy.
+type RetriesConfig struct {
+	MaxRetryAttempts int `mapstructure:"max_retry_attempts"`
+}
+
+// Set is part of config interface implementation.
+func (c *RetriesConfig) Set(dp config.DataProvider) error {
+	maxRetryAttempts, err := dp.GetInt(cfgKeyRetriesMax)
+	if err != nil {
+		return err
+	}
+	c.MaxRetryAttempts = maxRetryAttempts
+
+	return nil
+}
+
+// SetProviderDefaults is part of config interface implementation.
+func (c *RetriesConfig) SetProviderDefaults(dp config.DataProvider) {
+	dp.SetDefault(cfgKeyRetriesMax, DefaultMaxRetryAttempts)
+}
+
+// TransportOpts returns transport options.
+func (c *RetriesConfig) TransportOpts() RetryableRoundTripperOpts {
+	return RetryableRoundTripperOpts{MaxRetryAttempts: c.MaxRetryAttempts}
+}
+
+// Config represents options for HTTP client configuration.
+type Config struct {
```

Comment:
Please consider the following suggestions:

1) We usually store internal config structs [by value, not by pointer](https://github.com/acronis/go-appkit/blob/master/httpserver/config.go#L165-L168). I.e.:
```go
type Config struct {
	Retries       RetriesConfig   `mapstructure:"retries"`
	RateLimits RateLimitConfig `mapstructure:"rate_limits"`
	Timeout     time.Duration    `mapstructure:"timeout"`
}
```

2) All existing `Config` structs [have 2 constructors](https://github.com/acronis/go-appkit/blob/master/httpserver/config.go#L177-L185): `NewConfig` and `NewConfigWithKeyPrefix`. Let's do the same for the `httpclient.Config` too.

3) Let's add a compile-time check that `httpclient.Config` implements `config.Config` and `config.KeyPrefixProvider` interfaces.

```go
var _ config.Config = (*Config)(nil)
var _ config.KeyPrefixProvider = (*Config)(nil)
```